packages/core/src/oot/ocarina.c

```diff
@@ -145,8 +167,277 @@ void Ocarina_HandleLastPlayedSong(GameState_Play* play, Actor_Player* player, s1
+
+    /* Set the correct farore */
+    swapFarore();
+}
+
+static void HandleSongOfTime(GameState_Play* play)
+{
+    int msgState;
+
+    msgState = Message_GetState(&play->msgCtx);
+    if (msgState == 2)
+    {
+        /* Stop ocarina */
+        sInCustomSong = CUSTOM_SONG_NONE;
+
+        if (play->msgCtx.choice == 0)
+        {
+            ageSwap(play);
         }
+        else
+        {
+            PlaySound(0x480A);
+            play->msgCtx.ocarinaMode = 4; // OCARINA_MODE_END
+        }
+    }
+}
+
+static void PrepareSoaringScreen(GameState_Play* play)
+{
+    PauseContext* pauseCtx = &play->pauseCtx;
+    s16 i;
+
+    // Rumble_StateReset();
+
+    pauseCtx->switchPageTimer = 0;
+    pauseCtx->changing = 1; // PAUSE_MAIN_STATE_SWITCHING_PAGE;
+
+    // Set eye position and pageIndex such that scrolling left brings to the desired page
+    pauseCtx->eye.x = -64.0f; // sKaleidoSetupEyeX1[pauseCtx->pageIndex];
+    pauseCtx->eye.z = 0.0f; // sKaleidoSetupEyeZ1[pauseCtx->pageIndex];
+    pauseCtx->screen_idx = 1 ; // sKaleidoSetupKscpPos1[pauseCtx->pageIndex];
+    pauseCtx->infoPanelOffsetY = -40;
+
+    for (i = 0; i < ARRAY_SIZE(pauseCtx->worldMapPoints); i++) {
+        pauseCtx->worldMapPoints[i] = 0;
+    }
+
+    for (i = OWL_WARP_STONE_TOWER; i >= OWL_WARP_GREAT_BAY_COAST; i--) {
+        if ((gMmOwlFlags >> i) & 1) {
+            pauseCtx->worldMapPoints[i] = 1;
+            pauseCtx->cursorPoint[PAUSE_WORLD_MAP] = i;
+        }
+    }
+
+    if ((gMmOwlFlags >> 4) & 1) {
+        pauseCtx->cursorPoint[PAUSE_WORLD_MAP] = 4;
+    }
+}
+
+static void SetupSoaring(GameState_Play* play)
+{
+    PauseContext* pauseCtx = &play->pauseCtx;
+
+    if (play->interfaceCtx.restrictions.warpSongs)
+    {
+        PlayerDisplayTextBox(play, 0x88c, NULL);
+    }
+    else
+    {
+        // if in a dungeon or boss scene
+        // {
+        //      "Soar to Dungeon Entrance?"
+        // }
+        // else
+        if (!gMmOwlFlags)
+        {
+            PlayerDisplayTextBox(play, 0x88c, NULL);
+            soaringNoStatuesMessage(play);
+        }
+        else
+        {
+            gPrevPageIndex = pauseCtx->screen_idx;
+            gPrevCursorPoint = pauseCtx->cursorPoint[PAUSE_WORLD_MAP];
+            pauseCtx->screen_idx = 0;
+            pauseCtx->state = PAUSE_STATE_OWLWARP_0;
+            PrepareSoaringScreen(play);
+            pauseCtx->screen_idx = 1;
+            gSaveContext.prevHudVisibilityMode = gSaveContext.hudVisibilityMode;
+            gSaveContext.nextHudVisibilityMode = 0xC; // HUD_VISIBILITY_A | HUD_VISIBILITY_B;
+            R_UPDATE_RATE = 2;
+            if (Letterbox_GetSizeTarget() != 0) {
+                Letterbox_SetSizeTarget(0);
+            }
+            func_800C7200(1);
+            play->msgCtx.msgLength = 0;
+            play->msgCtx.msgMode = 0; // MSGMODE_NONE
+        }
+    }
+}
+
+EnTorch2* gElegyShells[1]; // need to increase this if other forms are added to OoT.
```

Comment:
No need for an array

Reply:
It's clearer this way that there can potentially be more shells in the future.

Reply:
We're not gonna put arrays everywhere just to handle hypothetical features.